kobo/apps/hook/tests/test_api_hook.py

```diff
@@ -0,0 +1,190 @@
+            "uid": hook.uid,
+        })
+
+        response = self.client.get(detail_url)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+        log_list_url = reverse("hook-log-list", kwargs={
+            "parent_lookup_asset": self.asset.uid,
+            "parent_lookup_hook": hook.uid,
+        })
+
+        response = self.client.get(log_list_url)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+
+    def test_create_hook(self):
+        self._create_hook()
+
+    def test_partial_update_hook(self):
+        hook = self._create_hook()
+        url = reverse("hook-detail", kwargs={
+            "parent_lookup_asset": self.asset.uid,
+            "uid": hook.uid
+        })
+        data = {
+            "name": "some disabled external service",
+            "active": False
+        }
+        response = self.client.patch(url, data, format=INSTANCE_FORMAT_TYPE_JSON)
+        self.assertEqual(response.status_code, status.HTTP_200_OK,
+                         msg=response.data)
+        hook.refresh_from_db()
+        self.assertFalse(hook.active)
+        self.assertEqual(hook.name, "some disabled external service")
+
+    @responses.activate
+    def test_send_and_retry(self):
+        hook = self._create_hook()
+
+        ServiceDefinition = hook.get_service_definition()
+        submissions = self.asset.deployment.get_submissions()
+        uuid = submissions[0].get("id")
+        service_definition = ServiceDefinition(hook, uuid)
+        first_mock_response = {"error": "not found"}
+
+        # Mock first requests try
+        responses.add(responses.POST, hook.endpoint,
+                      json=first_mock_response, status=status.HTTP_404_NOT_FOUND)
+        # Mock next requests tries
+        responses.add(responses.POST, hook.endpoint,
+                      status=status.HTTP_200_OK,
+                      content_type="application/json")
+
+        # Try to send data to external endpoint
+        success = service_definition.send()
+        self.assertFalse(success)
+
+        # Retrieve the corresponding log
+        url = reverse("hook-log-list", kwargs={
+            "parent_lookup_asset": hook.asset.uid,
+            "parent_lookup_hook": hook.uid
+        })
+
+        response = self.client.get(url, format=INSTANCE_FORMAT_TYPE_JSON)
+        first_hooklog = response.data.get("results")[0]
+
+        # Result should match first try
+        self.assertEqual(first_hooklog.get("status_code"), status.HTTP_404_NOT_FOUND)
+        self.assertEqual(json.loads(first_hooklog.get("message")), first_mock_response)
+
+        # Let's retry through API call
+        retry_url = reverse("hook-log-retry", kwargs={
+            "parent_lookup_asset": self.asset.uid,
+            "parent_lookup_hook": hook.uid,
+            "uid": first_hooklog.get("uid")
+        })
+
+        # Fakes Celery n retries by forcing status to `failed` (where n is `settings.HOOKLOG_MAX_RETRIES`)
```

Comment:
This confused me for a minute, but I understand it now to mean: the API will reject a retry request after only one failure, so manually set the status to HOOK_LOG_FAILED. This mimics what would happen after Celery has reached `settings.HOOKLOG_MAX_RETRIES`.

Reply:
Exactly!